docs/howto/features/auth-provider.md

```diff
@@ -0,0 +1,75 @@
+often just to use JupyterHub as an OAuth2 provider for user logins. This
+allows them to have the same shared users between the hub and their 
+service.
+
+2i2c can not support such external services, but we will provide
+appropriate credentials to them so they can do so themselves.
+
+## Setting up service credentials
+
+1. Define the service in the appropriate hub's `<hub-name>.values.yaml` file:
+
+   ```yaml
+   jupyterhub:
+     hub:
+       services:
+         # Name of the service should be provided to us by the community
+         <name-of-service>:
+           name: <name-of-service>
+           # The client_id for this service - it *must* start with service-
+           oauth_client_id: service-<name-of-service>
+           # The OAuth2 redirect URI provided to us by the community
+           oauth_redirect_uri: <oauth-redirect-uri>
+           # Set to true if it should be displayed under 'services' in the
+           # hub control panel
+           display: true/false
+           # Set this to false if there should be an extra 'do you want to
+           # approve this external service?' authorization screen. For external
+           # services, we mostly don't want it - so let's set this to true
+           oauth_no_confirm: true
+       config:
+         # This prevents the hub home page from showing when the end user
+         # tries to authenticate with the hub from the external service.
+         # Since the hub might send the user to CILogon or GitHub or another
+         # auth service, removing this step makes the flow smoother.
+         auto_login_oauth2_authorize: true
+   ```
+
+
+2. Generate an appropriate OAuth2 client secret on the commandline by running
+   `openssl rand -hex 32`
+   
+3. Define appropriate secrets in the hub's `enc-<hub-name>.secret.values.yaml` file:
+
+   ```yaml
+   jupyterhub:
+     hub:
+       services:
+         <name-of-service>:
+           # This will be the client secret
+           api_key: <output-of-step-2> 
+   ```
+           
+4. Commit and deploy this.
+
+5. Send our community the following pieces of information to configure their
+   OAuth2 *client* with:
+   
+   ```
+   client_id: <oauth_client_id_from_step_1>
+   client_secret: <api_key_from_step_3>
+   authorization_url: https://<hub-domain>/hub/api/oauth2/authorize
+   token_url: https://<hub-domain>/hub/api/oauth2/token
+   user_profile_url: https://<hub-domain>/hub/api/user
+   ```
+   
+   In addition, the following pieces of information would be useful:
+   
+   a. The `client_secret` **must** be sent as a token in the `Authorization` header when making the token request to the `token_url`.
+   b. The `name` field from the user profile contains the user's name on the hub.
+   
+   This should be enough information for them to configure their oauth2 client to authenticate
+   against the JupyterHub!
```

Comment:
It would be good if we can delegate some details to jupyterhub's docs:

- With regards to the authorization header, its documented in [how to make an api request](https://jupyterhub.readthedocs.io/en/stable/howto/rest.html#make-an-api-request), clarifying for example that you need a format of `Authorization: token <client_secret>`.
- There is also configuration references for all the endpoints
  - [GET `oauth2/authorize`](https://jupyterhub.readthedocs.io/en/stable/reference/rest-api.html#/default/get_oauth2_authorize)
  - [POST `oauth2/token`](https://jupyterhub.readthedocs.io/en/stable/reference/rest-api.html#/default/post_oauth2_token)
  - [GET `user`](https://jupyterhub.readthedocs.io/en/stable/reference/rest-api.html#/default/get_user)

  I've often looked at references for the userinfo endpoint as knowing what data is available and how its formatted is often very relevant and not standardized across identity providers.

Reply:
While in general I agree about delegation, I think in this case most people don't actually even know that OAuth2 requires 3 endpoints, so I think I would like to keep this here so a 2i2c engineer knows exactly what information to send to end user. I also forgot about the 3 endpoints while doing this with CosmicDS yesterday, and had to hunt around and look it up. Instead, this can be copy pasted and updated and sent to the community.

Reply:
I think by linking to configuration reference for the various endpoints are very relevant if we are also having a note about how to format the `Authorization` header.

In situations where I've setup authentication against a service, I've had great use of endpoint docs. In their shoes I may for example want to read the jupyterhub username from the userdata url - but what response should I expect and under what key should I look to find that?

I'll push a commit with a proposed resolution!